CMakeLists.txt

```diff
@@ -1,504 +1,2030 @@
-# CMAKE_C_COMPILER and CMAKE_CXX_COMPILER must be defined
-# before the project statement; However, the project
-# statement will clear CMAKE_BUILD_TYPE. CACHE variables,
-# along with the order of this code, are used to work
-# around these constraints.
-#
-# Don't put any code above or in this block, unless it
-# has similar constraints.
-cmake_minimum_required(VERSION 3.1.0)
-set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/Builds/CMake")
-include(CMakeFuncs)
-set(openssl_min 1.0.2)
-parse_target()
-project(rippled)
-#########################################################
-
-if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_BINARY_DIR}")
-  set(dir "build")
-  set(cmd "cmake")
-  if (target)
-    set(dir "${dir}/${target}")
-    set(cmd "${cmd} -Dtarget=${target}")
-  elseif(CMAKE_BUILD_TYPE)
-    set(dir "${dir}/${CMAKE_BUILD_TYPE}")
-    set(cmd "${cmd} -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}")
-  else()
-    set(dir "${dir}/default")
-  endif()
-  set(cmd "${cmd} ${CMAKE_SOURCE_DIR}")
-
-  message(FATAL_ERROR "Builds are not allowed in ${CMAKE_SOURCE_DIR}.\n"
-    "Instead:\n"
-    "1) Remove the CMakeCache.txt file and CMakeFiles directory "
-   "from ${CMAKE_SOURCE_DIR}.\n"
-    "2) Create a directory to hold your build files, for example: ${dir}.\n"
-    "3) Change to that directory.\n"
-    "4) Run cmake targetting ${CMAKE_SOURCE_DIR}, for example: ${cmd}")
-endif()
-if("${CMAKE_GENERATOR}" MATCHES "Visual Studio" AND
-    NOT ("${CMAKE_GENERATOR}" MATCHES .*Win64.*))
-  message(FATAL_ERROR "Visual Studio 32-bit build is unsupported. Use
-    -G\"${CMAKE_GENERATOR} Win64\"")
-endif()
-
-setup_build_cache()
-
-if(nonunity)
-    get_cmake_property(allvars VARIABLES)
-    string(REGEX MATCHALL "[^;]*(DEBUG|RELEASE)[^;]*" matchvars "${allvars}")
-    foreach(var IN LISTS matchvars)
-        string(REGEX REPLACE "(DEBUG|RELEASE)" "\\1CLASSIC" newvar ${var})
-        set(${newvar} ${${var}})
-    endforeach()
-
-    get_cmake_property(allvars CACHE_VARIABLES)
-    string(REGEX MATCHALL "[^;]*(DEBUG|RELEASE)[^;]*" matchvars "${allvars}")
-    foreach(var IN LISTS matchvars)
-        string(REGEX REPLACE "(DEBUG|RELEASE)" "\\1CLASSIC" newvar ${var})
-        set(${newvar} ${${var}} CACHE STRING "Copied from ${var}")
-    endforeach()
-endif()
-
-determine_build_type()
-
-check_gcc4_abi()
-
-############################################################
-
-include_directories(
-  src
-  src/beast
-  src/beast/extras
-  src/nudb/include
-  src/soci/src
-  src/soci/include)
-
-special_build_flags()
-
-############################################################
-
-use_boost(
-  # resist the temptation to alphabetize these. coroutine
-  # must come before context.
-  chrono
-  coroutine
-  context
-  date_time
-  filesystem
-  program_options
-  regex
-  serialization
-  system
-  thread)
-
-use_pthread()
-
-use_openssl(${openssl_min})
-
-use_protobuf()
 
-setup_build_boilerplate()
+cmake_minimum_required (VERSION 3.7.0)
```

Comment:
Was this version arbitrarily chosen or are you truly relying on features that only came in version 3.7.0?

Reply:
I guess I'm actually relying on 3.9 because of the protobuf find module (imported targets and static lib support):  https://cmake.org/cmake/help/latest/release/3.9.html.  I will update the required version here. I believe VS ships with some modified version of 3.9, so that's about as high as we can go right now I think.

Reply:
Makes sense 👍